File: src/java/foundgine-core/src/main/java/com/foundgine/core/semantic/mutation/SemanticMutationDependency.java
package com.foundgine.core.semantic.mutation;

import com.foundgine.core.abstractions.*;

/** Semantic dependency between mutation operations. */
public record SemanticMutationDependency(int sourceOperationIndex, int targetOperationIndex,
                                         FieldId sourceField, FieldId targetField, RelationshipId relationship) {
    public SemanticMutationDependency { if (sourceField == null || targetField == null) throw new NullPointerException(); }
    public SemanticMutationDependency(int sourceOperationIndex, int targetOperationIndex, FieldId sourceField, FieldId targetField) {
        this(sourceOperationIndex, targetOperationIndex, sourceField, targetField, null);
    }
}

File: src/csharp/Foundgine.Core/Semantic/Mutation/SemanticMutationDependency.cs
using Foundgine.Core.Abstractions;

namespace Foundgine.Core.Semantic.Mutation;

/// <summary>
/// A semantic dependency between mutation operations.
/// </summary>
public sealed record SemanticMutationDependency(
    int SourceOperationIndex,
    int TargetOperationIndex,
    FieldId SourceField,
    FieldId TargetField,
    RelationshipId? Relationship = null);
